From 2243cb621995024076328ab08020706623fe72b2 Mon Sep 17 00:00:00 2001
From: Oleg Karfich <oleg.karfich@wago.com>
Date: Fri, 22 Mar 2019 15:20:39 +0100
Subject: [PATCH] boards: add support for wago pac-768-330x boards

Since am335x based pac devices are equal to pfc v3 devices from
bootloader perspective, we integrate their support under the am335x-pfc
board folder.

Signed-off-by: Oleg Karfich <oleg.karfich@wago.com>
---
 arch/arm/boards/wago-pfc-am335x/board.c            |   1 +
 arch/arm/boards/wago-pfc-am335x/lowlevel.c         |   3 +
 arch/arm/configs/am335x_pac_defconfig              | 142 +++++++++++++++++++++
 arch/arm/configs/am335x_pac_mlo_defconfig          |  32 +++++
 arch/arm/dts/Makefile                              |   3 +
 arch/arm/dts/am335x-pac-768_330x-mlo.dts           |  50 ++++++++
 arch/arm/dts/am335x-pac-768_330x.dts               | 131 +++++++++++++++++++
 arch/arm/dts/am335x-pfc-leds.dtsi                  |  80 ++++++------
 arch/arm/dts/am335x-pfc.dtsi                       |  19 ++-
 arch/arm/mach-omap/Kconfig                         |   5 +
 .../defaultenv-pfc-base/init/a-pfc-switch-reset    |   3 +
 images/Makefile.am33xx                             |   8 ++
 12 files changed, 430 insertions(+), 47 deletions(-)
 create mode 100644 arch/arm/configs/am335x_pac_defconfig
 create mode 100644 arch/arm/configs/am335x_pac_mlo_defconfig
 create mode 100644 arch/arm/dts/am335x-pac-768_330x-mlo.dts
 create mode 100644 arch/arm/dts/am335x-pac-768_330x.dts

diff --git a/arch/arm/boards/wago-pfc-am335x/board.c b/arch/arm/boards/wago-pfc-am335x/board.c
index c857581..f088d72 100644
--- a/arch/arm/boards/wago-pfc-am335x/board.c
+++ b/arch/arm/boards/wago-pfc-am335x/board.c
@@ -8,6 +8,7 @@
  *   - 750-810x	PFC100
  *   - 750-8208	PFC200
  *   - 750-821x	PFC200
+ *   - 768-330x	PAC100
  *
  * This program is free software; you can redistribute it and/or
  * modify it under the terms of the GNU General Public License as
diff --git a/arch/arm/boards/wago-pfc-am335x/lowlevel.c b/arch/arm/boards/wago-pfc-am335x/lowlevel.c
index 5c5333d..79c3083 100644
--- a/arch/arm/boards/wago-pfc-am335x/lowlevel.c
+++ b/arch/arm/boards/wago-pfc-am335x/lowlevel.c
@@ -112,8 +112,11 @@ PFC_ENTRY_MLO(start_am33xx_pfc_750_821x_sram_512mb, am335x_pfc_750_821x_mlo,
 								PFC_DDR3_512MB);
 PFC_ENTRY_MLO(start_am33xx_src_2850_311x_sram_256mb, am335x_src_2850_311x_mlo,
 								PFC_DDR3_256MB);
+PFC_ENTRY_MLO(start_am33xx_pac_768_330x_sram_512mb, am335x_pac_768_330x_mlo,
+								PFC_DDR3_512MB);
 
 PFC_ENTRY(start_am33xx_pfc_750_810x_sdram, am335x_pfc_750_810x, SZ_256M);
 PFC_ENTRY(start_am33xx_pfc_750_8208_sdram, am335x_pfc_750_8208, SZ_256M);
 PFC_ENTRY(start_am33xx_pfc_750_821x_sdram, am335x_pfc_750_821x, SZ_512M);
 PFC_ENTRY(start_am33xx_src_2850_311x_sdram, am335x_src_2850_311x, SZ_256M);
+PFC_ENTRY(start_am33xx_pac_768_330x_sdram, am335x_pac_768_330x, SZ_512M);
diff --git a/arch/arm/configs/am335x_pac_defconfig b/arch/arm/configs/am335x_pac_defconfig
new file mode 100644
index 0000000..3fc8129
--- /dev/null
+++ b/arch/arm/configs/am335x_pac_defconfig
@@ -0,0 +1,142 @@
+CONFIG_ARCH_OMAP=y
+# CONFIG_OMAP_GPMC is not set
+CONFIG_BAREBOX_UPDATE_AM33XX_EMMC=y
+CONFIG_OMAP_MULTI_BOARDS=y
+CONFIG_MACH_WAGO_PFC_AM335X=y
+CONFIG_MACH_WAGO_PFC_AM335X_768_330x=y
+CONFIG_THUMB2_BAREBOX=y
+CONFIG_ARM_BOARD_APPEND_ATAG=y
+CONFIG_ARM_OPTIMZED_STRING_FUNCTIONS=y
+CONFIG_ARM_UNWIND=y
+CONFIG_MMU=y
+CONFIG_MALLOC_SIZE=0x0
+CONFIG_MALLOC_TLSF=y
+CONFIG_KALLSYMS=y
+CONFIG_RELOCATABLE=y
+CONFIG_PROMPT="barebox> "
+CONFIG_HUSH_FANCY_PROMPT=y
+CONFIG_AUTO_COMPLETE=y
+CONFIG_MENU=y
+# CONFIG_TIMESTAMP is not set
+CONFIG_BOOTM_SHOW_TYPE=y
+CONFIG_BOOTM_VERBOSE=y
+CONFIG_BOOTM_INITRD=y
+CONFIG_BOOTM_OFTREE=y
+CONFIG_BLSPEC=y
+CONFIG_CONSOLE_ACTIVATE_NONE=y
+CONFIG_CONSOLE_ALLOW_COLOR=y
+CONFIG_CONSOLE_RATP=y
+CONFIG_DEFAULT_ENVIRONMENT_GENERIC_NEW=y
+CONFIG_DEFAULT_ENVIRONMENT_PATH="defaultenv/defaultenv-pfc-base/ arch/arm/boards/wago-pfc-am335x/defaultenv-pfc-am335x/"
+CONFIG_STATE=y
+CONFIG_BOOTCHOOSER=y
+CONFIG_RESET_SOURCE=y
+CONFIG_DEBUG_INFO=y
+CONFIG_LONGHELP=y
+CONFIG_CMD_IOMEM=y
+CONFIG_CMD_MEMINFO=y
+CONFIG_CMD_ARM_MMUINFO=y
+CONFIG_CMD_MMC_EXTCSD=y
+# CONFIG_CMD_BOOTU is not set
+CONFIG_CMD_BOOTZ=y
+CONFIG_CMD_GO=y
+CONFIG_CMD_LOADB=y
+CONFIG_CMD_RESET=y
+CONFIG_CMD_UIMAGE=y
+CONFIG_CMD_PARTITION=y
+CONFIG_CMD_EXPORT=y
+CONFIG_CMD_LOADENV=y
+CONFIG_CMD_PRINTENV=y
+CONFIG_CMD_MAGICVAR=y
+CONFIG_CMD_MAGICVAR_HELP=y
+CONFIG_CMD_SAVEENV=y
+CONFIG_CMD_FILETYPE=y
+CONFIG_CMD_LN=y
+CONFIG_CMD_MD5SUM=y
+CONFIG_CMD_UNCOMPRESS=y
+CONFIG_CMD_LET=y
+CONFIG_CMD_MSLEEP=y
+CONFIG_CMD_READF=y
+CONFIG_CMD_SLEEP=y
+CONFIG_CMD_DHCP=y
+CONFIG_CMD_MIITOOL=y
+CONFIG_CMD_PING=y
+CONFIG_CMD_TFTP=y
+CONFIG_CMD_ECHO_E=y
+CONFIG_CMD_EDIT=y
+CONFIG_CMD_MENU=y
+CONFIG_CMD_MENUTREE=y
+CONFIG_CMD_READLINE=y
+CONFIG_CMD_TIMEOUT=y
+CONFIG_CMD_CRC=y
+CONFIG_CMD_CRC_CMP=y
+CONFIG_CMD_MM=y
+CONFIG_CMD_DETECT=y
+CONFIG_CMD_FLASH=y
+CONFIG_CMD_GPIO=y
+CONFIG_CMD_I2C=y
+CONFIG_CMD_LED=y
+CONFIG_CMD_LED_MODE=y
+CONFIG_CMD_LED_TRIGGER=y
+CONFIG_CMD_USBGADGET=y
+CONFIG_CMD_WD=y
+CONFIG_CMD_BAREBOX_UPDATE=y
+CONFIG_CMD_OF_NODE=y
+CONFIG_CMD_OF_PROPERTY=y
+CONFIG_CMD_OF_DISPLAY_TIMINGS=y
+CONFIG_CMD_OF_FIXUP_STATUS=y
+CONFIG_CMD_OFTREE=y
+CONFIG_CMD_TIME=y
+CONFIG_CMD_STATE=y
+CONFIG_CMD_BOOTCHOOSER=y
+CONFIG_CMD_UART_CONSOLE=y
+CONFIG_CMD_ETHADDR_MATCH=y
+CONFIG_NET=y
+CONFIG_NET_NFS=y
+CONFIG_NET_NETCONSOLE=y
+CONFIG_NET_RESOLV=y
+CONFIG_OF_BAREBOX_DRIVERS=y
+CONFIG_OF_BAREBOX_ENV_IN_FS=y
+CONFIG_DRIVER_SERIAL_NS16550=y
+CONFIG_DRIVER_SERIAL_NS16550_OMAP_EXTENSIONS=y
+CONFIG_DRIVER_NET_CPSW=y
+CONFIG_MICREL_PHY=y
+CONFIG_SMSC_PHY=y
+CONFIG_NET_USB=y
+CONFIG_NET_USB_ASIX=y
+CONFIG_NET_USB_SMSC95XX=y
+# CONFIG_SPI is not set
+CONFIG_MTD=y
+CONFIG_USB_HOST=y
+CONFIG_USB_STORAGE=y
+CONFIG_USB_GADGET=y
+CONFIG_USB_GADGET_DFU=y
+CONFIG_USB_GADGET_SERIAL=y
+CONFIG_USB_GADGET_FASTBOOT=y
+CONFIG_USB_MUSB=y
+CONFIG_USB_MUSB_AM335X=y
+CONFIG_USB_MUSB_HOST=y
+CONFIG_USB_MUSB_GADGET=y
+CONFIG_MCI=y
+CONFIG_MCI_STARTUP=y
+CONFIG_MCI_OMAP_HSMMC=y
+CONFIG_STATE_DRV=y
+CONFIG_XSECTION=y
+CONFIG_BOOTMATRIX=y
+CONFIG_LED=y
+CONFIG_LED_GPIO=y
+CONFIG_LED_GPIO_OF=y
+CONFIG_LED_TRIGGERS=y
+CONFIG_LED_PCA955X=y
+CONFIG_WATCHDOG=y
+CONFIG_WATCHDOG_OMAP=y
+CONFIG_GPIO_GENERIC_PLATFORM=y
+CONFIG_GPIO_PCA953X=y
+CONFIG_PINCTRL_SINGLE=y
+CONFIG_FS_EXT4=y
+CONFIG_FS_TFTP=y
+CONFIG_FS_NFS=y
+CONFIG_FS_FAT=y
+CONFIG_FS_FAT_WRITE=y
+CONFIG_FS_FAT_LFN=y
+CONFIG_ZLIB=y
diff --git a/arch/arm/configs/am335x_pac_mlo_defconfig b/arch/arm/configs/am335x_pac_mlo_defconfig
new file mode 100644
index 0000000..c90c3ab
--- /dev/null
+++ b/arch/arm/configs/am335x_pac_mlo_defconfig
@@ -0,0 +1,32 @@
+CONFIG_ARCH_OMAP=y
+# CONFIG_OMAP_GPMC is not set
+CONFIG_OMAP_BUILD_IFT=y
+CONFIG_OMAP_MULTI_BOARDS=y
+CONFIG_MACH_WAGO_PFC_AM335X=y
+CONFIG_MACH_WAGO_PFC_AM335X_768_330x=y
+CONFIG_THUMB2_BAREBOX=y
+CONFIG_MMU=y
+CONFIG_BAREBOX_MAX_PBLX_SIZE=0x1b400
+CONFIG_MALLOC_SIZE=0x0
+CONFIG_MALLOC_TLSF=y
+CONFIG_RELOCATABLE=y
+CONFIG_PROMPT="MLO>"
+CONFIG_SHELL_NONE=y
+# CONFIG_ERRNO_MESSAGES is not set
+# CONFIG_TIMESTAMP is not set
+CONFIG_DEBUG_INFO=y
+CONFIG_OFDEVICE=y
+CONFIG_DRIVER_SERIAL_NS16550=y
+CONFIG_DRIVER_SERIAL_NS16550_OMAP_EXTENSIONS=y
+# CONFIG_SPI is not set
+CONFIG_I2C=y
+CONFIG_I2C_OMAP=y
+CONFIG_MCI=y
+CONFIG_MCI_STARTUP=y
+# CONFIG_MCI_WRITE is not set
+CONFIG_MCI_OMAP_HSMMC=y
+CONFIG_EEPROM_AT24=y
+CONFIG_PINCTRL_SINGLE=y
+# CONFIG_FS_DEVFS is not set
+CONFIG_FS_FAT=y
+CONFIG_FS_FAT_LFN=y
diff --git a/arch/arm/dts/Makefile b/arch/arm/dts/Makefile
index 2fbbe0f..38cf36d 100644
--- a/arch/arm/dts/Makefile
+++ b/arch/arm/dts/Makefile
@@ -127,4 +127,7 @@ pbl-dtb-$(CONFIG_MACH_WAGO_PFC_AM335X_2850_311x) += \
 	am335x-src-2850_311x-mlo.dtb.o
 pbl-dtb-$(CONFIG_MACH_WAGO_PFC_AM35XX) += am35xx-pfc-750_820x.dtb.o
 pbl-dtb-$(CONFIG_MACH_WAGO_VTPCTP_MX6) += imx6q-vtpctp-762-4xxx.dtb.o
+pbl-dtb-$(CONFIG_MACH_WAGO_PFC_AM335X_768_330x) += \
+	am335x-pac-768_330x.dtb.o \
+	am335x-pac-768_330x-mlo.dtb.o
 clean-files := *.dtb *.dtb.S .*.dtc .*.pre .*.dts *.dtb.lzo
diff --git a/arch/arm/dts/am335x-pac-768_330x-mlo.dts b/arch/arm/dts/am335x-pac-768_330x-mlo.dts
new file mode 100644
index 0000000..4ca8d31
--- /dev/null
+++ b/arch/arm/dts/am335x-pac-768_330x-mlo.dts
@@ -0,0 +1,50 @@
+/*
+ * Copyright (C) 2017 WAGO Kontakttechnik GmbH & Co. KG <http://global.wago.com>
+ * Author: Heinrich Toews <heinrich.toews@wago.com>
+ *
+ * Copyright (C) 2018 WAGO Kontakttechnik GmbH & Co. KG <http://global.wago.com>
+ * Author: Oleg Karfich <oleg.karfich@wago.com>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of
+ * the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+/dts-v1/;
+
+#include "am335x-pfc.dtsi"
+#include "am335x-pfc-emmc.dtsi"
+#include "am335x-pfc-strip.dtsi"
+#include "am33xx-strip.dtsi"
+#include "am33xx-clocks-strip.dtsi"
+
+/ {
+	model = "PAC-768-330x MLO";
+	compatible = "wago,am335x-pac-768_330x", "wago,am335x-pfc", "ti,am33xx";
+	board-variant = "pac100";
+
+	memory {
+		device_type = "memory";
+		reg = <0x80000000 0x20000000>; /* 512 MB */
+	};
+};
+
+&mmc1_cd_pins {
+	pinctrl-single,pins = <
+		0x04C (PIN_INPUT_PULLDOWN | MUX_MODE7)  /* gpmc_a3.gpio1_19 */
+	>;
+};
+
+&mmc1 {
+	cd-gpios = <&gpio1 19 GPIO_ACTIVE_HIGH>;
+};
+
+&mmc2 {
+	status = "okay";
+};
diff --git a/arch/arm/dts/am335x-pac-768_330x.dts b/arch/arm/dts/am335x-pac-768_330x.dts
new file mode 100644
index 0000000..f381a07
--- /dev/null
+++ b/arch/arm/dts/am335x-pac-768_330x.dts
@@ -0,0 +1,131 @@
+/*
+ * Copyright (C) 2017 WAGO Kontakttechnik GmbH & Co. KG <http://global.wago.com>
+ * Author: Heinrich Toews <heinrich.toews@wago.com>
+ *
+ * Copyright (C) 2018 WAGO Kontakttechnik GmbH & Co. KG <http://global.wago.com>
+ * Author: Oleg Karfich <oleg.karfich@wago.com>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of
+ * the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+/dts-v1/;
+
+#include "am335x-pfc.dtsi"
+#include "am335x-pfc-emmc.dtsi"
+#include "am335x-pfc-leds.dtsi"
+#include "am335x-pfc-rmii12.dtsi"
+
+/ {
+	model = "PAC-768-330x";
+	compatible = "wago,am335x-pac-768_330x", "wago,am335x-pfc", "ti,am33xx";
+	board-variant = "pac100";
+
+	memory {
+		device_type = "memory";
+		reg = <0x80000000 0x20000000>; /* 512 MB */
+	};
+};
+
+&i2c0 {
+	pca953x: pca953x@70 {
+		compatible = "nxp,pca9538";
+		reg = <0x70>;
+		gpio-controller;
+		#gpio-cells = <2>;
+	};
+};
+
+&mmc1_cd_pins {
+	pinctrl-single,pins = <
+		0x04C (PIN_INPUT_PULLDOWN | MUX_MODE7)  /* gpmc_a3.gpio1_19 */
+	>;
+};
+
+&mmc1 {
+	cd-gpios = <&gpio1 19 GPIO_ACTIVE_HIGH>;
+};
+
+&mmc2 {
+	status = "okay";
+};
+
+&pca9552_60 {
+	status = "okay";
+};
+
+&led_0 {
+	label = "ds-red";
+};
+
+&led_1 {
+	label = "ds-green";
+};
+
+&led_2 {
+	label = "ec-red";
+};
+
+&led_3 {
+	label = "ec-blue";
+};
+
+&led_4 {
+	label = "ms-red";
+};
+
+&led_5 {
+	label = "ms-green";
+};
+
+&led_6 {
+	label = "ns-red";
+};
+
+&led_7 {
+	label = "ns-green";
+};
+
+&led_8 {
+	label = "run-red";
+};
+
+&led_9 {
+	label = "run-green";
+};
+
+&led_10 {
+	label = "app-red";
+};
+
+&led_11 {
+	label = "app-green";
+};
+
+&mac {
+	status = "okay";
+};
+
+&usb {
+	status = "okay";
+};
+
+&usb_ctrl_mod {
+	status = "okay";
+};
+
+&usb0 {
+	dr_mode = "otg";
+	status = "okay";
+};
+
+&usb0_phy {
+	status = "okay";
+};
diff --git a/arch/arm/dts/am335x-pfc-leds.dtsi b/arch/arm/dts/am335x-pfc-leds.dtsi
index 97c58b8..fe90719 100644
--- a/arch/arm/dts/am335x-pfc-leds.dtsi
+++ b/arch/arm/dts/am335x-pfc-leds.dtsi
@@ -125,82 +125,82 @@
 		reg = <0x60>;
 		status = "disabled";
 
-		overwrite@15 {
+		overwrite: overwrite@15 {
 			label = "overwrite";
 			reg = <15>;
 			linux,default-trigger = "default-on";
 		};
 
-		sys-red@0 {
+		led_0: sys-red@0 {
 			label = "sys-red";
 			reg = <0>;
 			linux,default-trigger = "none";
 		};
 
-		sys-green@1 {
+		led_1: sys-green@1 {
 			label = "sys-green";
 			reg = <1>;
 			linux,default-trigger = "none";
 		};
 
-		run-red@2 {
+		led_2: run-red@2 {
 			label = "run-red";
 			reg = <2>;
 			linux,default-trigger = "none";
 		};
 
-		run-green@3 {
-		      label = "run-green";
-		      reg = <3>;
-		      linux,default-trigger = "none";
+		led_3: run-green@3 {
+			label = "run-green";
+			reg = <3>;
+			linux,default-trigger = "none";
 		};
 
-		io-red@4 {
-		      label = "io-red";
-		      reg = <4>;
-		      linux,default-trigger = "none";
+		led_4: io-red@4 {
+			label = "io-red";
+			reg = <4>;
+			linux,default-trigger = "none";
 		};
 
-		io-green@5 {
-		      label = "io-green";
-		      reg = <5>;
-		      linux,default-trigger = "none";
+		led_5: io-green@5 {
+			label = "io-green";
+			reg = <5>;
+			linux,default-trigger = "none";
 		};
 
-		ms-red@6 {
-		      label = "ms-red";
-		      reg = <6>;
-		      linux,default-trigger = "none";
+		led_6: ms-red@6 {
+			label = "ms-red";
+			reg = <6>;
+			linux,default-trigger = "none";
 		};
 
-		ms-green@7 {
-		      label = "ms-green";
-		      reg = <7>;
-		      linux,default-trigger = "none";
+		led_7: ms-green@7 {
+			label = "ms-green";
+			reg = <7>;
+			linux,default-trigger = "none";
 		};
 
-		ns-red@8 {
-		      label = "ns-red";
-		      reg = <8>;
-		      linux,default-trigger = "none";
+		led_8: ns-red@8 {
+			label = "ns-red";
+			reg = <8>;
+			linux,default-trigger = "none";
 		};
 
-		ns-green@9 {
-		      label = "ns-green";
-		      reg = <9>;
-		      linux,default-trigger = "none";
+		led_9: ns-green@9 {
+			label = "ns-green";
+			reg = <9>;
+			linux,default-trigger = "none";
 		};
 
-		u7-red@10 {
-		      label = "u7-red";
-		      reg = <10>;
-		      linux,default-trigger = "none";
+		led_10: u7-red@10 {
+			label = "u7-red";
+			reg = <10>;
+			linux,default-trigger = "none";
 		};
 
-		u7-green@11 {
-		      label = "u7-green";
-		      reg = <11>;
-		      linux,default-trigger = "none";
+		led_11: u7-green@11 {
+			label = "u7-green";
+			reg = <11>;
+			linux,default-trigger = "none";
 		};
 	};
 };
diff --git a/arch/arm/dts/am335x-pfc.dtsi b/arch/arm/dts/am335x-pfc.dtsi
index 2a39a2e..c88361a 100644
--- a/arch/arm/dts/am335x-pfc.dtsi
+++ b/arch/arm/dts/am335x-pfc.dtsi
@@ -126,12 +126,17 @@
 
 	mmc1_pins: pinmux_mmc1_pins {
 		pinctrl-single,pins = <
-                        0x0F0 (PIN_INPUT_PULLUP | MUX_MODE0)	/* mmc0_dat3.mmc0_dat3 */
-                        0x0F4 (PIN_INPUT_PULLUP | MUX_MODE0)	/* mmc0_dat2.mmc0_dat2 */
-                        0x0F8 (PIN_INPUT_PULLUP | MUX_MODE0)	/* mmc0_dat1.mmc0_dat1 */
-                        0x0FC (PIN_INPUT_PULLUP | MUX_MODE0)	/* mmc0_dat0.mmc0_dat0 */
-                        0x100 (PIN_INPUT_PULLUP | MUX_MODE0)	/* mmc0_clk.mmc0_clk   */
-                        0x104 (PIN_INPUT_PULLUP | MUX_MODE0)	/* mmc0_cmd.mmc0_cmd   */
+			0x0F0 (PIN_INPUT_PULLUP | MUX_MODE0)	/* mmc0_dat3.mmc0_dat3 */
+			0x0F4 (PIN_INPUT_PULLUP | MUX_MODE0)	/* mmc0_dat2.mmc0_dat2 */
+			0x0F8 (PIN_INPUT_PULLUP | MUX_MODE0)	/* mmc0_dat1.mmc0_dat1 */
+			0x0FC (PIN_INPUT_PULLUP | MUX_MODE0)	/* mmc0_dat0.mmc0_dat0 */
+			0x100 (PIN_INPUT_PULLUP | MUX_MODE0)	/* mmc0_clk.mmc0_clk   */
+			0x104 (PIN_INPUT_PULLUP | MUX_MODE0)	/* mmc0_cmd.mmc0_cmd   */
+		>;
+	};
+
+	mmc1_cd_pins: pinmux_mmc1_cd_pins {
+		pinctrl-single,pins = <
 			0x190 (PIN_INPUT_PULLDOWN | MUX_MODE7)  /* mcasp0_aclkx.gpio3_14 */
 		>;
 	};
@@ -155,7 +160,7 @@
 	vmmc-supply = <&vmmcsd_fixed>;
 	bus-width = <0x4>;
 	pinctrl-names = "default";
-	pinctrl-0 = <&mmc1_pins>;
+	pinctrl-0 = <&mmc1_pins>, <&mmc1_cd_pins>;
 	cd-gpios = <&gpio3 14 GPIO_ACTIVE_HIGH>;
 	cd-inverted;
 };
diff --git a/arch/arm/mach-omap/Kconfig b/arch/arm/mach-omap/Kconfig
index 81a9e96..9c93007 100644
--- a/arch/arm/mach-omap/Kconfig
+++ b/arch/arm/mach-omap/Kconfig
@@ -213,6 +213,11 @@ config MACH_WAGO_PFC_AM335X_2850_311x
 	help
 	  Say Y here if you are using a am335x based Wago SRC-2850-311x
 
+config MACH_WAGO_PFC_AM335X_768_330x
+	bool "PAC-768-330x"
+	help
+	  Say Y here if you are using a am335x based Wago PAC-768-330x
+
 endif
 
 endif
diff --git a/defaultenv/defaultenv-pfc-base/init/a-pfc-switch-reset b/defaultenv/defaultenv-pfc-base/init/a-pfc-switch-reset
index 3dc8de5..0d33a3d 100644
--- a/defaultenv/defaultenv-pfc-base/init/a-pfc-switch-reset
+++ b/defaultenv/defaultenv-pfc-base/init/a-pfc-switch-reset
@@ -10,6 +10,7 @@ pfc200v2_sw_reset_pin=106
 pfc200v3_sw_reset_pin=132
 vtpctp_sw_reset_pin=105
 src_sw_reset_pin=${pfc200v2_sw_reset_pin}
+pac100_sw_reset_pin=${pfc200v3_sw_reset_pin}
 
 if [ "${global.board.variant}" = "pfc100" ]; then
 	current_reset_pin="${pfc100_sw_reset_pin}"
@@ -21,6 +22,8 @@ elif [ "${global.board.variant}" = "vtpctp" ]; then
 	current_reset_pin="${vtpctp_sw_reset_pin}"
 elif [ "${global.board.variant}" = "src" ]; then
 	current_reset_pin="${src_sw_reset_pin}"
+elif [ "${global.board.variant}" = "pac100" ]; then
+	current_reset_pin="${pac100_sw_reset_pin}"
 else
 	echo -e ${warning} "no switch reset pin defined for board \e[00;32m${global.board.variant}\e[00m"
 	exit 0
diff --git a/images/Makefile.am33xx b/images/Makefile.am33xx
index cead7d6..f78defe 100644
--- a/images/Makefile.am33xx
+++ b/images/Makefile.am33xx
@@ -167,6 +167,14 @@ pblx-$(CONFIG_MACH_WAGO_PFC_AM335X_2850_311x) += start_am33xx_src_2850_311x_sram
 FILE_barebox-am33xx-src-2850-311x-mlo.img = start_am33xx_src_2850_311x_sram_256mb.pblx.mlo
 am33xx-mlo-$(CONFIG_MACH_WAGO_PFC_AM335X_2850_311x) += barebox-am33xx-src-2850-311x-mlo.img
 
+pblx-$(CONFIG_MACH_WAGO_PFC_AM335X_768_330x) += start_am33xx_pac_768_330x_sdram
+FILE_barebox-am33xx-pac-768-330x.img = start_am33xx_pac_768_330x_sdram.pblx
+am33xx-barebox-$(CONFIG_MACH_WAGO_PFC_AM335X_768_330x) += barebox-am33xx-pac-768-330x.img
+
+pblx-$(CONFIG_MACH_WAGO_PFC_AM335X_768_330x) += start_am33xx_pac_768_330x_sram_512mb
+FILE_barebox-am33xx-pac-768-330x-mlo.img = start_am33xx_pac_768_330x_sram_512mb.pblx.mlo
+am33xx-mlo-$(CONFIG_MACH_WAGO_PFC_AM335X_768_330x) += barebox-am33xx-pac-768-330x-mlo.img
+
 ifdef CONFIG_OMAP_BUILD_IFT
 image-y += $(am33xx-mlo-y)
 else
-- 
2.7.4

